### PR TITLE
Moonshine logits processor

### PR DESCRIPTION
**Pull request by eustlb**
# What does this PR do?

This PR introduces a new logits processor that enables per-batch-index maximum lengths.

This feature is particularly useful for audio encoder-decoder models that rely on generate and use a heuristic based on a token_limit_factor. In such cases, the maximum length can be inferred from the input length, given a token/sec maximum. This approach helps reduce model hallucinations.

The logits processor is designed to be passed via the logits_processor argument of the generate method. Additionally, this PR leverages the logits processor in the Moonshine model, where per-batch-index maximum lengths are inferred from the provided inputs.

Besides, I also fixed typos in the docs and updated tests for T4 runners.

### CODE REVIEW

File: src/transformers/models/moonshine/modular_moonshine.py [hunk 1]
```diff
@@ -1109,22 +1098,106 @@ def forward(
...
+    def generate(
+        self,
+        input_values: Optional[torch.FloatTensor] = None,
+        infer_max_length_from_input: bool = True,
+        tokens_per_second: float = None,
+        **generate_kwargs,
+    ):
+        """
+        Overrides GenerationMixin's [`~generation.GenerationMixin.generate`] method to add a `PerBatchIndexMaxLengthLogitsProcessor` by default.
+        Indeed Moonshine model uses per-batch-index max length that are inferred from the input audio length.
```
Comment (stevhliu) [inline src/transformers/models/moonshine/modular_moonshine.py:1110]:
```suggestion
        Overrides GenerationMixin's [`~generation.GenerationMixin.generate`] method to add a `PerBatchIndexMaxLengthLogitsProcessor` as default because Moonshine uses per-batch-index max length that is inferred from the input audio length.
```
```

File: docs/source/en/_toctree.yml [hunk 2]
```diff
@@ -772,6 +770,8 @@
         title: Mimi
       - local: model_doc/mms
         title: MMS
+      - local: model_doc/moonshine
+        title: moonshine
```
Comment (stevhliu) [inline docs/source/en/_toctree.yml:774]:
```suggestion
        title: Moonshine
```

**Review by stevhliu — APPROVED** [review]
Very nice, thanks!

I think it'd also be helpful to make a mention of this in the "Tips" section of `moonshine.md`! It's kinda easy to forget that this feature exists when it is enabled by default

**Review by gante — COMMENTED** [review]
@eustlb From our discussion on Slack I was not understanding the purpose of the new logits processor, but now I get it :D `moonshine` accepts variable-length audio input, we know that audio has roughly X tokens per second, and thus we want to help the model avoid the hallucination region through the maximum generated length. 

With padded inputs, different inputs will have different durations, and therefore we would need to control the maximum admissible length of each batch member individually 💡 

Suggestions/ideas: 
1. Instead of a new logits processor to control the maximum length with a new argument (`max_lengths`), what do you think of expanding `max_length` and `MaxLengthCriteria` to accept one input per row? That way a) this feature would also become easily available to other models b) no new arguments would be added
2. [incompatible with the suggestion above, and perhaps better for this model] The technique you implemented effectively sets a hard cutoff -- at X length, end generation through an eos token, regardless of the confidence of the model. Usually, it's more effective to set a soft cutoff, i.e. gradually increase the probability of the eos token after a certain length, as in `ExponentialDecayLengthPenalty`. With a soft cutoff, you can set it a few tokens earlier than before, and it becomes almost a hard cutoff after a few tokens, with the difference that high-confidence generation would almost never be cut too early. Have you considered/experimented with these techniques?

**Comment by eustlb** [comment]
Thanks a lot @gante for your insights! 🐐

My idea with this first draft was to modify the existing logic as little as possible. However, I definitely considered option 1, and I’m glad you suggested it. Indeed, such an approach aligns, in my opinion, with the direction of standardizing audio processing, as it could benefit other models as well. Additionally, implementing it directly in `MaxLengthCriteria` makes the solution cleaner (and less hacky), as termination would stem from the stopping criteria rather than being forced by the logits processor. This approach aligns better with the behavior of `generate`.

That said, this would require `generate` to support `max_length` and `max_new_tokens` as lists (and similarly for `min_length` and `min_new_tokens`). Consequently, `generation_config.max_length`, etc., would also need to be able to be lists within the scope of generate. This is inconsistent with how these attributes are currently defined in [GenerationConfig](https://github.com/huggingface/transformers/blob/615bf9c5e4ab29ff63be8d5ba49525d6e3d1b3a7/src/transformers/generation/configuration_utils.py#L98) (and there's no reason to change them). **Are we okay with this that?**

If we are, I can update `MaxLengthCriteria` to accept a list of integers. This way, Moonshine’s processor can simply return `max_length` as a list, inferred by default from the provided audio inputs, and there would be no need to override generate 😊

**Comment by gante** [comment]
> This is inconsistent with how these attributes are currently defined in [GenerationConfig](https://github.com/huggingface/transformers/blob/615bf9c5e4ab29ff63be8d5ba49525d6e3d1b3a7/src/transformers/generation/configuration_utils.py#L98) (and there's no reason to change them). Are we okay with this that?

IMO since it's a relevant feature for time-based output modalities like audio (and likely video in the near-future), I'd be happy to support it.

BTW, beam search + stopping criteria is not working at the moment 💔 , see [this issue](https://github.com/huggingface/transformers/issues/34843). This may be a deal breaker is beam search is relevant for moonshine, in which case I would recommend the current form.

**Comment by eustlb** [comment]
Following this discussion, I've split this PR in two PRs:
1. https://github.com/huggingface/transformers/pull/35731 patch moonshine
2. https://github.com/huggingface/transformers/pull/35732 generate's with expanded `max_length` etc
